velox/docs/develop/aggregate-functions.rst

```diff
@@ -109,25 +109,323 @@ accordingly.  Data is stored in the following order:
+
+When `T` is Varchar, Varbinary, or a complex type, `exec::arg_type<T>`,
+`exec::optional_arg_type<T>`, and `exec::out_type<T>` are the corresponding
+view and writer types of `T`. A datailed explanaion can be found in :doc:`view-and-writer-types`.
+
+.. list-table::
+   :widths: 25 25
+   :header-rows: 1
+
+   * - Default-Null Behavior
+     - Non-Default-Null Behavior
+   * - static bool SimpleAverageAggregate::toIntermediate(
+          exec::out_type<Row<double, int64_t>>& out,
+          exec::arg_type<T> in);
+     - static bool SimpleArrayAggAggregate::toIntermediate(
+          exec::out_type<Array<Generic<T1>>>& out,
+          exec::optional_arg_type<Generic<T1>> in);
+
+AccumulatorType of Default-Null Behavior
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For aggregaiton functions of default-null behavior, the author defines an
+`AccumulatorType` struct as follows.
+
+.. code-block:: c++
+
+  struct AccumulatorType {
+    // Author defines data members
+    ...
+
+    // Optional. Default is true.
+    static constexpr bool is_fixed_size_ = false;
+
+    // Optional. Default is false.
+    static constexpr bool use_external_memory_ = true;
+
+    explicit AccumulatorType(HashStringAllocator* allocator);
+
+    void addInput(HashStringAllocator* allocator, exec::arg_type<T1> value1, ...);
+
+    void combine(
+        HashStringAllocator* allocator,
+        exec::arg_type<IntermediateType> other);
+
+    bool writeIntermediateResult(exec::out_type<IntermediateType>& out);
+
+    bool writeFinalResult(exec::out_type<OutputType>& out);
+
+    // Optional. Called during destruction.
+    void destroy(HashStringAllocator* allocator);
+  };
+
+The author defines an optional flag `is_fixed_size_` indicating whether the
+every accumulator takes fixed amount of memory. This flag is true by default.
+Next, the author defines another optional flag `use_external_memory_`
+indicating whether the accumulator uses memory that is not tracked by Velox.
+This flag is false by default.
+
+The author defines a constructor that takes a single argument of
+`HashStringAllocator*`. This constructor is called before aggregation starts to
+initialize all accumulators.
+
+The author can also optionally define a `destroy` function that is called when
+*this* accumulator object is destructed.
+
+addInput
+""""""""
+
+This method adds raw input values to *this* accumulator. It receives a
+`HashStringAllocator*` followed by `exec::arg_type<T1>`-typed values, one for
+each argument type `Ti` wrapped in InputType.
+
+By default-null behavior, raw-input rows where at least one column is null are
+ignored before `addInput` is called. After `addInput` is called, *this*
+accumulator is assumed to be non-null.
+
+combine
+"""""""
+
+This method adds an input intermediate state to *this* accumulator. It receives
+a `HashStringAllocator*` and one `exec::arg_type<IntermediateType>` value. By
+default-null behavior, nulls among the input intermediate states are ignore
```

Comment:
nit; With default-null behavior, nulls among the input intermediate states are ignored before combine is called.